### PR TITLE
Fix duplicate work in pmatch' impl of builtin lists

### PR DESCRIPTION
**Pull request by TotallyNotChase**
This also tries to set up a test for this kind of sanity tests.

### CODE REVIEW

File: examples/Examples/PlutusType.hs [hunk 1]
```diff
@@ -39,13 +40,50 @@ swap x = pmatch x $ \case
...
+  `traces` replicate 1 tag
+#else
+  `traces` []
+#endif
+  where
+    tag = "evaluating"
+
+-- FIXME: This needs to fully "evaluate" (e.g evaluate in UPLC) the `x` within the `pmatch` handler.
+-- | Make sure the target of 'pmatch' is only evaluated once.
+_pmatchTargetEval :: HasTester => PlutusType p => ClosedTerm p -> Assertion
```
Comment (TotallyNotChase) [inline examples/Examples/PlutusType.hs:81]: This is what I was trying. It's missing one key piece- the argument to the `pmatch` handler needs to have its inner terms "evaluated" with `plet` (see: `pmatchTargetEvalBList` impl). Only then this test is valid and can work generically on any plutus type.

The thing is, I don't think we can do that without another typeclass.

Comment (TotallyNotChase) [reply examples/Examples/PlutusType.hs]: Actually never mind, I think I fixed it.

**Review by L-as — APPROVED** [review]
LGTM

File: examples/Examples/PlutusType.hs [hunk 2]
```diff
@@ -39,13 +40,36 @@ swap x = pmatch x $ \case
...
+-- CPP support isn't great in fourmolu.
+{- ORMOLU_DISABLE -}
+
+-- | Make sure the target of 'pmatch' is only evaluated once.
+pmatchTargetEval :: HasTester => PlutusType p => ClosedTerm p -> Assertion
+pmatchTargetEval target = pmatch (ptrace (pconstant tag) target) (\x -> plet (pcon x) $ \_ -> pconstant ())
+#ifdef Development
+  `traces` replicate 1 tag
+#else
+  `traces` []
```
[the remark this answers is not in the document]
Comment (TotallyNotChase) [reply examples/Examples/PlutusType.hs:72]: Good idea. I'll make a note to make that change the next time I need to touch the test suite - unless it's already taken care of.

**Pull request merged**